### PR TITLE
Implementation of CGPattern via runtime base

### LINKED ISSUE #2353: [Foundation Dependency Removal] CGPattern should be implemented via CFRuntimeBase

### PR DESCRIPTION
**Pull request by msft-Jeyaram**
fixes #2353

### CODE REVIEW

File: Frameworks/CoreGraphics/CGPattern.mm [hunk 1]
```diff
@@ -17,160 +17,181 @@
...
-                                       kCGImageAlphaPremultipliedFirst | kCGBitmapByteOrder32Big);
-    CGRect rect = { 0, 0, width, height };
-    CGContextDrawTiledImage(patternCtx, rect, tilePattern);
-
-    CGColorSpaceRelease(colorSpace);
-    CGImageRelease(tilePattern);
-
-    if (generatedImage) {
-        CGImageRelease(generatedImage);
+    inline CGImageRef GetPatternImage() {
```
Comment (msft-Jeyaram) [inline Frameworks/CoreGraphics/CGPattern.mm]: this isn't doing it's job properly and sadly the code isn't exercised properly.
I have a fix for this via CGColor<->UIColor work.

Note: This CR is mainly to address moving CGPattern off of NSObject and having it implemented via runtime base + minor bug fixes.

Comment (msft-Jeyaram) [inline Frameworks/CoreGraphics/CGPattern.mm]: hmm I might as well do it here.

---
In reply to: [108832278](https://github.com/Microsoft/WinObjC/pull/2374#discussion_r108832278) [](ancestors = 108832278)

File: Frameworks/UIKit/UIColor.mm [hunk 2]
```diff
@@ -518,12 +518,12 @@ - (UIColor*)initWithPatternImage:(UIImage*)image {
 callbacks.releaseInfo = 0;
 callbacks.drawPattern = __UIColorPatternFill;
 
-_pattern = (id) CGPatternCreateColorspace(self, bounds, m, bounds.size.width, bounds.size.height, 0, NO, &callbacks, pImg->_has32BitAlpha ? _ColorABGR : _ColorBGR);
+_pattern = _CGPatternCreateColorspace(self, bounds, m, bounds.size.width, bounds.size.height, 0, NO, &callbacks, pImg->_has32BitAlpha ? _ColorABGR : _ColorBGR);
```
Comment (rajsesh-msft) [inline Frameworks/UIKit/UIColor.mm]: Lets remove the code in #if 0.

File: Frameworks/UIKit/UIColor.mm [hunk 3]
```diff
@@ -224,7 +224,7 @@ @implementation UIColor {
 @public
     enum BrushType _type;
     UIImage* _image;
```
Comment (rajsesh-msft) [inline Frameworks/UIKit/UIColor.mm]: nit: while we are at this, could this be StrongId<>?

File: Frameworks/CoreGraphics/CGPattern.mm [hunk 4]
```diff
@@ -17,160 +17,98 @@
...
-    return matrix;
-}
+    inline void IssuePatternCallback(CGContextRef context) const {
+        _callbacks.drawPattern(_info, context);
+    }
 
-- (void)dealloc {
-    if (generatedImage) {
-        CGImageRelease(generatedImage);
+    inline CGAffineTransform TransformMatrix() const {
```
Comment (rajsesh-msft) [inline Frameworks/CoreGraphics/CGPattern.mm]: Nit: The name sounds like a verb.  Adding a Get will clarify the intent.

File: Frameworks/CoreGraphics/CGPattern.mm [hunk 5]
```diff
@@ -17,160 +17,98 @@
...
-                                       kCGImageAlphaPremultipliedFirst | kCGBitmapByteOrder32Big);
-    CGRect rect = { 0, 0, width, height };
-    CGContextDrawTiledImage(patternCtx, rect, tilePattern);
-
-    CGColorSpaceRelease(colorSpace);
-    CGImageRelease(tilePattern);
-
-    if (generatedImage) {
-        CGImageRelease(generatedImage);
+    inline CGRect Bounds() const {
```
Comment (rajsesh-msft) [inline Frameworks/CoreGraphics/CGPattern.mm]: nit: could return const ref.

File: Frameworks/CoreGraphics/CGPattern.mm [hunk 6]
```diff
@@ -17,160 +17,98 @@
...
-    outHeight += height;
-    }*/
-
-    if (isColored) {
-        bitmapInfo = kCGImageAlphaFirst | kCGBitmapByteOrder32Big;
-        colorSpace = CGColorSpaceCreateDeviceRGB();
-    } else {
-        bitmapInfo = kCGImageAlphaNone;
-        colorSpace = CGColorSpaceCreateDeviceGray();
+    inline CGAffineTransform PatternTransform() {
```
Comment (rajsesh-msft) [inline Frameworks/CoreGraphics/CGPattern.mm]: nit: GetPatternTransform() const

File: Frameworks/CoreGraphics/CGPattern.mm [hunk 7]
```diff
@@ -17,160 +17,98 @@
...
 
-    // TODO: Caching.
-    generatedImage = CGBitmapContextCreateImage(patternCtx);
-    CGImageRetain(generatedImage);
-
-    CGContextRelease(patternCtx);
-
-    return generatedImage;
-}
+    inline CGPatternCallbacks Callbacks() const {
```
Comment (DHowett-MSFT) [inline Frameworks/CoreGraphics/CGPattern.mm]: This is unused. Either use it or delete `IssuePatternCallback`

File: Frameworks/include/CGPatternInternal.h [hunk 8]
```diff
@@ -58,5 +42,14 @@ CGRect _CGPatternGetFinalPatternSize(CGPatternRef pattern);
 */
 bool _CGPatternIsColored(CGPatternRef pattern);
 
+COREGRAPHICS_EXPORT CGPatternRef _CGPatternCreateColorspace(void* info,
```
Comment (DHowett-MSFT) [inline Frameworks/include/CGPatternInternal.h]: this doesn't actually take or create a color space?

File: build/CoreGraphics/dll/CoreGraphics.def [hunk 9]
```diff
@@ -450,6 +450,7 @@ LIBRARY CoreGraphics
 
         ; private exports below
         _CGPatternCreateFromImage
+        _CGPatternCreateColorspace
```
Comment (DHowett-MSFT) [inline build/CoreGraphics/dll/CoreGraphics.def]: this is only used internally, why is it exported?

Comment (msft-Jeyaram) [inline build/CoreGraphics/dll/CoreGraphics.def]: actually it was in UIColor but now that's being removed. taking it out.

---
In reply to: [109017192](https://github.com/Microsoft/WinObjC/pull/2374#discussion_r109017192) [](ancestors = 109017192)

File: Frameworks/include/CGPatternInternal.h (hunk 8 above)
Comment (msft-Jeyaram) [inline Frameworks/include/CGPatternInternal.h]: no longer needed.

---
In reply to: [109017103](https://github.com/Microsoft/WinObjC/pull/2374#discussion_r109017103) [](ancestors = 109017103)

**Comment by msft-Jeyaram** [comment]
@DHowett-MSFT @rajsesh-msft  updated :)

File: Frameworks/CoreGraphics/CGPattern.mm [hunk 10]
```diff
@@ -17,160 +17,78 @@
...
-+ (instancetype)patternWithImage:(CGImageRef)img {
-    CGPattern* ret = [CGPattern new];
-
-    ret->matrix = CGAffineTransformIdentity;
-    ret->simpleImage = img;
-    CFRetain((id)img);
-
-    return ret;
-}
+static const wchar_t* TAG = L"CGPattern";
```
Comment (MSFTFox) [inline Frameworks/CoreGraphics/CGPattern.mm]: Seems unused with LoggingNative?

File: Frameworks/CoreGraphics/CGPattern.mm [hunk 11]
```diff
@@ -17,160 +17,78 @@
...
-    return ret;
-}
+
+private:
+    woc::StrongCF<CGImageRef> _generatedImage;
+    woc::StrongCF<CGImageRef> _simpleImage;
+    void* _info;
+    CGRect _bounds;
+    CGAffineTransform _transformMatrix;
+    CGFloat _xStep;
```
Comment (MSFTFox) [inline Frameworks/CoreGraphics/CGPattern.mm]: xstep and ystep could be CGSize for simplicity.

Comment (msft-Jeyaram) [inline Frameworks/CoreGraphics/CGPattern.mm]: it's much more implicit this way.

---
In reply to: [109041128](https://github.com/Microsoft/WinObjC/pull/2374#discussion_r109041128) [](ancestors = 109041128)

**Comment by msft-Jeyaram** [comment]
Awaiting ARM test verification.

**Pull request merged**